Workflow Execution Logic

BlakeSLC opened the pull request.

BlakeSLC committed:
initial commit with workflow execution

BlakeSLC committed:
updates to nuget & more

BlakeSLC committed:
remove log

BlakeSLC committed:
minor efficiency improvement

ArneMaes0 committed:
Removed the ConnectorAPI references for QActions that do not need it.

ArneMaes0 commented on QAction_1/InterApp/Executors/Workflows/WorkflowValidation.cs:
```diff
@@ -2,180 +2,209 @@
+                return false;
+            }
+
+            // Check if the public keys are fetched
+            if (repo.PublicKey == Exceptions.NotAvailable ||
+                repo.PublicKeyID == Exceptions.NotAvailable)
+            {
+                result = $"The public keys are not available for '{workflowRequest.RepositoryId.FullName}'. Either the configured API Token does not have permission to the repository, or the public keys for the repository are not fetched yet.";
+                return false;
+            }
```
Do you need this check for the execute workflow request? The public keys are used to create new repository secrets, but are they needed for the execution as well?

BlakeSLC replied:
Removed, I figured all requests needed them but double checking the API it seems not

ArneMaes0 committed:
Changed the triggers and actions to be consistent with the others
Also added the change history I forgot last release

ArneMaes0 commented on QAction_1/InterApp/Executors/Workflows/ExecuteWorkflowExecutor.cs:
```diff
@@ -0,0 +1,106 @@
+            protocol.CheckTrigger(231);
+
+            // Return message
+            result = null;
```
I would do the actual building of the request in the RepositoriesRequestHandler class as all the request are build there and it can be reused elsewhere in the connector too.

ArneMaes0 commented on QAction_1/InterApp/Executors/Workflows/ExecuteWorkflowExecutor.cs:
```diff
@@ -0,0 +1,106 @@
+            {
+                @ref = Message.Data.WorkflowReference,
+                inputs = Message.Data.WorkflowInputs,
+            };
```
Could be a class as well instead of an anonymous one. It's then consistent in the way the other requests are build

ArneMaes0 commented on QAction_1/PollManager/RequestType.cs:
```diff
@@ -66,7 +66,10 @@ public enum RequestType
-	}
+
+        [Description("Workflow Execution")]
+        Workflow_Execution = 81,
```
Make the naming consistent with the other requests.

ArneMaes0 commented on QAction_1/PollManager/ResponseHandler/Repositories/WorkflowsResponseHandler.cs:
```diff
@@ -2,132 +2,175 @@
+            var urlParts = executionUrl.Split('/');
+            var owner = urlParts[1];
+            var repoName = urlParts[2];
+            var workflowId = urlParts[urlParts.Length - 2];
```
Maybe do the regex way so it's consistent with the rest of the driver. It's then also more readable for newcomers so they see what the actual url looks like

ArneMaes0 commented on QAction_1/PollManager/ResponseHandler/Repositories/WorkflowsResponseHandler.cs:
```diff
@@ -2,132 +2,175 @@
+    using Skyline.Protocol.PollManager.RequestHandler.Repositories;
+    using Skyline.Protocol.Tables;
+
+    public static class WorkflowResponseHandler
```
I would've keep the partial RepositoriesResponseHandler class as it's triggering a workflow for a specific repository

ArneMaes0 commented on QAction_1/PollManager/ResponseHandler/Repositories/WorkflowsResponseHandler.cs:
```diff
@@ -2,132 +2,175 @@
+            // Parse response
+            var parameters = (object[])protocol.GetParameters(new uint[] { Parameter.getrepositoryworkflowscontent_205, Parameter.getrepositoryworkflowsurl_105 });
+            var response = JsonConvert.DeserializeObject<RepositoryWorkflowsResponse>(Convert.ToString(parameters[0]));
+            var url = Convert.ToString(parameters[1]);
```
Is this speed improvement worth the readability and breaking consistency across reponse handler for?

BlakeSLC replied:
a bit confused as I think this code is not something I touched, maybe because I copied it to a new handler?

ArneMaes0 replied:
No problem my bad, it showed up as a change in the VS comparison. I've just reverted it so it's consistent with the rest of the connector.

ArneMaes0 committed:
CR 1

ArneMaes0 commented on protocol.xml:
```diff
@@ -602,7 +602,11 @@ To get more request you can add an api key under General/Configuration page. To
+			<Name>ProcessWorkflowExecution</Name>
+			<Description>Process Workflow Execution</Description>
+			<Type>dummy</Type>
+		</Param>
```
The rest of the connector does the QAction triggering based on the response content parameter. This could be removed and the QAction20001 trigger could be updated with the Param 231

BlakeSLC replied:
that didn't work, as the response content is consistently empty. A trigger after group is required
https://docs.github.com/en/rest/actions/workflows?apiVersion=2022-11-28#create-a-workflow-dispatch-event

the only response is a status code 204

ArneMaes0 replied:
Aah yes indeed, maybe we can then reuse the param 231 which is normally the response content but since your request doesn't have any response content we can make that one a dummy to trigger the handler QAction.

ArneMaes0 replied:
Thanks for letting me know! I've reused the link param (PID 281) as the link header is not used in the dispatch response anyway. And also added parsing of the body (the response only has a body in case of a failure) and use the error message in the interapp return message. For instance if you try to trigger a workflow that isn't manually triggerable the Github Api returns this response:
```json
{
    "message": "Workflow does not have 'workflow_dispatch' trigger",
    "documentation_url": "https://docs.github.com/rest/actions/workflows#create-a-workflow-dispatch-event",
    "status": "422"
}
```

which I found a usefull thing to return in the InterApp message.

ArneMaes0 commented on protocol.xml:
```diff
@@ -5325,7 +5329,7 @@ Are you sure you want to continue.]]>
 
 		<QAction id="20000" name="HTTP Request Handling" encoding="csharp" triggers="30" />
-		<QAction id="20001" name="HTTP Parse Responses" encoding="csharp" triggers="201;202;203;204;205;210;211;212;213;220;221;222;223;224;226;228;229;230;231"></QAction>
+		<QAction id="20001" name="HTTP Parse Responses" encoding="csharp" triggers="201;202;203;204;205;210;211;212;213;220;221;222;223;224;226;228;229;230;81"></QAction>
```
Same as the other remark on Param 81

ArneMaes0 commented on protocol.xml:
```diff
@@ -6291,6 +6295,15 @@ Are you sure you want to continue.]]>
+			<Content>
+				<Id>281</Id>
+			</Content>
+		</Trigger>
```
Do the clearing of the link header param as this is used in all the sessions

ArneMaes0 commented on protocol.xml:
```diff
@@ -6497,6 +6510,11 @@ Are you sure you want to continue.]]>
+			<Name>Process Workflow Execution</Name>
+			<On id="81">parameter</On>
+			<Type>run actions</Type>
+		</Action>
```
All the other Actions in the 250+ range are all clear actions, for consistancy I would make this one the clear param for the link header. And then also add clear action for the response content with id 331

ArneMaes0 commented:
I implemented most of my remarks on your pr as they were small things

ArneMaes0 commented:
The only things that are still open are the:

1. The public key validation step
2. The minor speed improvement change

BlakeSLC committed:
remarks adjusted

BlakeSLC committed:
re-add in after group trigger logic

BlakeSLC commented:
> fixed only a couple things, as I think you did most of them. However, one change you made will break the functionality (removing the trigger after group to run the QAction). I've added that back in as a separate commit, so we can discuss and adjust if needed

after re-reading now I understand, but honestly think the functionality is a bit less efficient. I'm guessing clear action after trigger will allow proper execution of the QAction, even if the response is 'blank'. I tried looking at the documentation but I don't know if clearing the parameter will run the QAction or not. If it doesn't, then we can do it your way. but if clearing a parameter runs the associated QAction, I feel this way is inefficient

BlakeSLC committed:
minor message update on bad workflow execution so driver doesn't get stuck

ArneMaes0 committed:
Reused the link header param for the workflow execute dummy trigger param as this is not used for the dispatch request
Parse the response content of the dispatch event (this only happens if the request fails) and use the message in the interapp response.

ArneMaes0 committed:
Used the secure deserializer for the workflow response handler

ArneMaes0 commented:
Great work Blake! I'll merge everything in and release new versions.

ArneMaes0 commented:
And just to add to the clearing triggering the QAction, I tried it and clearing does not trigger the QAction so it didn't work using my way.

ArneMaes0 committed:
DIS Protocol Utils suggestion

ArneMaes0 committed:
Used the new official Githu Repositories ConnectorAPI v1.0.7
